File: excercise/ejercicio.cpp
#include <bits/stdc++.h>
#include <chrono> // Para medir el tiempo de ejecución
using namespace std;

struct Item {
    int profit, weight;
    Item(int profit, int weight) : profit(profit), weight(weight) {}
};

static bool cmp(struct Item a, struct Item b) {
    double r1 = (double)a.profit / (double)a.weight;
    double r2 = (double)b.profit / (double)b.weight;
    return r1 > r2;
}

double fractionalKnapsack(int W, struct Item arr[], int N) {
    sort(arr, arr + N, cmp);
    double finalvalue = 0.0;
    for (int i = 0; i < N; i++) {
        if (arr[i].weight <= W) {
            W -= arr[i].weight;
            finalvalue += arr[i].profit;
        } else {
            finalvalue += arr[i].profit * ((double)W / (double)arr[i].weight);
            break;
        }
    }
    return finalvalue;
}

int main() {
    Item arr[] = { 
        { 100, 10 }, { 280, 40 }, { 120, 20 }, { 120, 24 }, { 100, 36 },
        { 150, 30 }, { 180, 50 }, { 40, 10 }, { 60, 20 }, { 90, 35 },
        { 200, 25 }, { 300, 29 }, { 90, 24 }, { 40, 5 }, { 30, 6 },
        { 10, 4 }, { 70, 20 }, { 80, 10 }, { 50, 15 }, { 110, 22 }
    };
    int N = sizeof(arr) / sizeof(arr[0]);

    int weights[] = { 50, 80 };
    for (int W : weights) {
        auto start = chrono::high_resolution_clock::now();
        double maxProfit = fractionalKnapsack(W, arr, N);
        
        auto end = chrono::high_resolution_clock::now();
        chrono::duration<double> duration = end - start;
        cout << "Para W = " << W << ", máximo beneficio = " << maxProfit 
             << ", tiempo de ejecución = " << duration.count() << " segundos" << endl;
    }
    return 0;
}

File: excercise/ejercicio.py
from time import time

class Item:
    def __init__(self, profit, weight):
        self.profit = profit
        self.weight = weight

def cmp(item):
    # Retorna la relación beneficio/peso para ordenar los artículos
    return item.profit / item.weight

def fractional_knapsack(W, items):
    # Ordena los artículos por beneficio/peso en orden descendente
    items.sort(key=cmp, reverse=True)
    
    final_value = 0.0
    for item in items:
        if item.weight <= W:
            W -= item.weight
            final_value += item.profit
        else:
            final_value += item.profit * (W / item.weight)
            break
    return final_value

# Lista de artículos
items = [
    Item(100, 10), Item(280, 40), Item(120, 20), Item(120, 24), Item(100, 36),
    Item(150, 30), Item(180, 50), Item(40, 10), Item(60, 20), Item(90, 35),
    Item(200, 25), Item(300, 29), Item(90, 24), Item(40, 5), Item(30, 6),
    Item(10, 4), Item(70, 20), Item(80, 10), Item(50, 15), Item(110, 22)
]

# Capacidades de la mochila
capacities = [50, 80]

for W in capacities:
    start = time()
    max_profit = fractional_knapsack(W, items)
    end = time()
    duration = end - start
    print(f"Para W = {W}, máximo beneficio = {max_profit:.2f}, tiempo de ejecución = {duration:.6f} segundos")
